source/dotnet/Library/AdaptiveCards/AdaptiveTypedElementConverter.cs

```diff
@@ -66,84 +55,115 @@ public static void RegisterTypedElement<T>(string typeName = null)
                         throw new AdaptiveSerializationException($"Required property 'id' not found on '{typeName}'");
                     }
                 }
-                else
+
+                // add id of element to ParseContext
+                AdaptiveInternalID internalID = AdaptiveInternalID.Current();
+                if (type != typeof(AdaptiveCard))
                 {
-                    string objectId = jObject.Value<string>("id");
-                    if (Ids.Contains(objectId))
-                    {
-                        throw new AdaptiveSerializationException($"Duplicate 'id' found: '{objectId}'");
-                    }
-                    else
-                    {
-                        Ids.Add(objectId);
-                    }
+                    internalID = AdaptiveInternalID.Next();
+                    ParseContext.PushElement(objectId, internalID);
                 }
 
                 var result = (AdaptiveTypedElement)Activator.CreateInstance(type);
                 try
                 {
                     serializer.Populate(jObject.CreateReader(), result);
                 }
-                catch (JsonSerializationException)
+                catch (JsonSerializationException) { }
+
+                // remove id of element from ParseContext
+                if (type != typeof(AdaptiveCard))
+                    ParseContext.PopElement();
+
+                HandleAdditionalProperties(result);
+                return result;
+            }
+            else // We're looking at an unknown element
+            {
+                string objectId = jObject.Value<string>("id");
+                AdaptiveInternalID internalID = AdaptiveInternalID.Next();
+
+                // Handle deserializing unknown element
+                ParseContext.PushElement(objectId, internalID);
+                AdaptiveTypedElement result = null;
+                if (ParseContext.Type == ParseContext.ContextType.Element)
+                {
+                    try
+                    {
+                        result = (AdaptiveTypedElement)Activator.CreateInstance(typeof(AdaptiveUnknownElement));
+                        serializer.Populate(jObject.CreateReader(), result);
+                    }
+                    catch (JsonSerializationException) { }
+                }
+                else // ParseContext.Type == ParseContext.ContextType.Action
                 {
-                    return result;
+                    try
+                    {
+                        result = (AdaptiveTypedElement)Activator.CreateInstance(typeof(AdaptiveUnknownAction));
+                        serializer.Populate(jObject.CreateReader(), result);
+                    }
+                    catch (JsonSerializationException) { }
                 }
+                ParseContext.PopElement();
 
-                HandleAdditionalProperties(result);
+                Warnings.Add(new AdaptiveWarning(-1, $"Unknown element '{typeName}'"));
```

Comment:
>1 [](start = 50, length = 1)

What's the -1 signify?

Reply:
we require an enum int here. unfortunately there's not a centralized one available. I've filed #2479 for this.

---
In reply to: [278323851](https://github.com/Microsoft/AdaptiveCards/pull/2560#discussion_r278323851) [](ancestors = 278323851)